src/modm/platform/can/stm32-fdcan/message_ram.hpp

```diff
@@ -0,0 +1,270 @@
+		(ExtendedFilterCount * ExtendedFilterSize)    +
+		(RxFifoElements      * RxFifoElementSize) * 2 +
+		(TxEventFifoEntries  * TxEventFifoEntrySize)  +
+		(TxFifoElements      * TxFifoElementSize);
+
+	static constexpr uintptr_t RamBase = SRAMCAN_BASE + (Size * InstanceIndex);
+
+	static constexpr uintptr_t FilterListStandard = RamBase + 0;
+	static constexpr uintptr_t FilterListExtended = FilterListStandard + (StandardFilterCount * StandardFilterSize);
+	static constexpr uintptr_t RxFifo0 = FilterListExtended + (ExtendedFilterCount * ExtendedFilterSize);
+	static constexpr uintptr_t RxFifo1 = RxFifo0 + (RxFifoElements * RxFifoElementSize);
+	static constexpr uintptr_t TxEventFifo = RxFifo1 + (RxFifoElements * RxFifoElementSize);
+	static constexpr uintptr_t TxFifo = TxEventFifo + (TxEventFifoEntries * TxEventFifoEntrySize);
+
+	/// Address of element in RX FIFO
+	struct RxFifoAddress
+	{
+		uint8_t fifoIndex;
+		uint8_t getIndex;
+
+		/// \returns pointer to RX fifo element
+		uint32_t* ptr()
+		{
+			uintptr_t address = RxFifo0 +
+				RxFifoElementSize * ((RxFifoElements * uint8_t(fifoIndex)) + getIndex);
+			return reinterpret_cast<uint32_t*>(address);
+		}
+	};
+
+	/// \returns pointer to element in TX queue
+	static uint32_t*
+	txFifoElement(uint8_t putIndex)
+	{
+		return reinterpret_cast<uint32_t*>(TxFifo + (putIndex * TxFifoElementSize));
+	}
+
+	/// \returns pointer to standard filter element
+	static uint32_t*
+	standardFilter(uint8_t index)
+	{
+		const auto address = FilterListStandard + (index * StandardFilterSize);
+		return reinterpret_cast<uint32_t*>(address);
+	}
+
+	/// \returns pointer to extended filter element
+	static uint32_t*
+	extendedFilter(uint8_t index)
+	{
+		const auto address = FilterListExtended + (index * ExtendedFilterSize);
+		return reinterpret_cast<uint32_t*>(address);
+	}
+public:
+	/// Write TX element headers to TX queue
+	static void
+	writeTxHeaders(uint8_t putIndex, CommonFifoHeader_t common, TxFifoHeader_t tx)
+	{
+		uint32_t* messageRam = txFifoElement(putIndex);
+		*messageRam++ = common.value;
+		*messageRam = tx.value;
+	}
+
+	/// Read RX element headers from RX fifo
+	static std::tuple<CommonFifoHeader, RxFifoHeader>
+	readRxHeaders(RxFifoAddress address)
+	{
+		const uint32_t* messageRam = address.ptr();
+		const auto commonHeader = CommonFifoHeader{*messageRam++};
+		const auto rxHeader = RxFifoHeader{*messageRam};
+		return {commonHeader, rxHeader};
+	}
+
+	/// Read message data from RX fifo
+	static void
+	readData(RxFifoAddress address, std::span<uint8_t> outData)
+	{
+		const auto size = std::min(outData.size(), 64u);
+
+		// + 2: skip 2x32 bit headers
+		const uint32_t* messageRam = address.ptr() + 2;
+		// message data must be read in 32bit accesses, memcpy on message ram does not work
+		for (auto i = 0u; i < size; i += 4) {
+			const uint32_t receivedData = *messageRam++;
+
+			const auto byteCount = std::min(4u, size - i);
+			std::memcpy(&outData[i], &receivedData, byteCount);
```

Comment:
I don't like this double copying of the message data.
Is it somehow possible to enforce 32bit alignment of `outData`?

Or the _Unaligned Memory Accessor_ could be used: https://docs.modm.io/develop/api/nucleo-g474re/group__modm__architecture__unaligned.html

Reply:
Casting any `uint8_t` array pointer to `uint32_t*` and dereferencing the pointer is a strict aliasing violation and therefore UB. This is a separate issue on top of alignment. You coudl get by with volatile or inline assembly.

What you do anyway first is a `LDR` instruction to load from memory to a register. Then you store from there to memory again, even if you would assign the data directly to a `uint32_t` location in memory (see https://gcc.godbolt.org/z/595KE5qzr)
memcpy could be replaced with something else because the code-gen has a function call, but I am not sure the improvement would be that big.

Reply:
I looked at `isUnaligned`, this is undefined behaviour, except for the M0 version. On Cortex M4 it will just do a reinterpret_cast and dereference the pointer. The M0 version accesses the data as `uint8_t` and copies to an internal `uint8_t` temporary buffer which is legal but not helpful since we could directly copy the data to the target.

Reply:
If we assume the destination buffer size is big enough to fit the full multiples of 4, we can change `byteCount` to be 4. Since we control the can message type, this is true. Then code generation is optimal (1 load, 1 store): https://gcc.godbolt.org/z/eaj4bejjz

Reply:
AFAIK the behavior is pretty well defined in GCC for ARM, Cortex-M3/4 (but not Cortex-M0) can do unaligned access in hardware.

But I don't care that much, so just keep it as it is for now. Maybe later with long-frames and DMA involved find a better solution.

Reply:
It's undefined behaviour and not related to hardware, but the language itself, see for example:
- https://blog.regehr.org/archives/959
- https://gist.github.com/shafik/848ae25ee209f698763cffee272a58f8
- https://cellperformance.beyond3d.com/articles/2006/06/understanding-strict-aliasing.html
- https://en.cppreference.com/w/cpp/language/reinterpret_cast#Type_aliasing

GCC only has an exception for accessing the inactive member of a union which would also be undefined behaviour.

I will change the size in memcpy to 4. Then the compiler emits just one `str` for memcpy. This will probably generate optimal assembly. It's perfectly fine as long as the can message buffer fits the full multiples of 4 bytes.